File: data_structures/adjacency_list_graph/python/adjacency_list_graph.py
"""隣接リストグラフ（有向グラフ）

操作:
  add_vertex(v)       : O(1)
  add_edge(u, v)      : O(1)
  has_edge(u, v)      : O(deg(u))
  get_neighbors(v)    : O(1)
  vertex_count()      : O(1)
  edge_count()        : O(1)
"""

from typing import Dict, List


from typing import Generic, TypeVar

V = TypeVar("V")


class AdjacencyListGraph(Generic[V]):
    def __init__(self) -> None:
        self._adj: Dict[V, List[V]] = {}
        self._edge_count = 0

    def add_vertex(self, v: V) -> None:
        if v not in self._adj:
            self._adj[v] = []

    def add_edge(self, u: V, v: V) -> None:
        self.add_vertex(u)
        self.add_vertex(v)
        self._adj[u].append(v)
        self._edge_count += 1

    def has_edge(self, u: V, v: V) -> bool:
        return u in self._adj and v in self._adj[u]

    def get_neighbors(self, v: V) -> List[V]:
        if v not in self._adj:
            raise KeyError(f"vertex {v} not found")
        return self._adj[v]

    def has_vertex(self, v: V) -> bool:
        return v in self._adj

    def vertex_count(self) -> int:
        return len(self._adj)

    def edge_count(self) -> int:
        return self._edge_count

File: data_structures/adjacency_list_graph/cpp/adjacency_list_graph.cpp
#include "adjacency_list_graph.h"
#include <algorithm>

void AdjacencyListGraph::add_vertex(int v) {
    adj_.emplace(v, std::vector<int>{});
}

void AdjacencyListGraph::add_edge(int u, int v) {
    if (!has_vertex(u)) add_vertex(u);
    if (!has_vertex(v)) add_vertex(v);
    adj_[u].push_back(v);
    ++edge_count_;
}

bool AdjacencyListGraph::has_edge(int u, int v) const {
    auto it = adj_.find(u);
    if (it == adj_.end()) return false;
    return std::find(it->second.begin(), it->second.end(), v) != it->second.end();
}

const std::vector<int>& AdjacencyListGraph::get_neighbors(int v) const {
    auto it = adj_.find(v);
    if (it == adj_.end()) throw std::out_of_range("vertex not found");
    return it->second;
}

bool AdjacencyListGraph::has_vertex(int v) const { return adj_.count(v) > 0; }
int AdjacencyListGraph::vertex_count() const { return static_cast<int>(adj_.size()); }
int AdjacencyListGraph::edge_count() const { return edge_count_; }
